src/vs/workbench/contrib/files/browser/views/explorerDecorationsProvider.ts

```diff
@@ -65,10 +65,10 @@ export class ExplorerDecorationsProvider implements IDecorationsProvider {
 		return this._onDidChange.event;
 	}
 
-	provideDecorations(resource: URI): IDecorationData | undefined {
+	async provideDecorations(resource: URI): Promise<IDecorationData | undefined> {
```

Comment:
I do not understand the purpose of this change. Why make it async? And throw?

Reply:
If we return `undefined` then `decorationsService` will assume it has no decorations at all for this provider and it won't try to fetch them again, so I just make it a promise and reject it to trigger [this logic](https://github.com/microsoft/vscode/blob/ad5ef7ea85b8a622f73b96394507d98809e68861/src/vs/workbench/services/decorations/browser/decorationsService.ts#L389-L391) and decorations can be fetched when explorerItem is resolved

Reply:
Throwing an error is a little fishy... Can't this use the `onDidChange` event and trigger a refresh that way?

Reply:
> Can't this use the onDidChange event and trigger a refresh that way?

There isn't an event that fires when an [ExplorerItem is resolved](https://github.com/microsoft/vscode/blob/75a6ddc8626b831f01f91f9f60f27de3d5bb6f38/src/vs/workbench/contrib/files/browser/views/explorerViewer.ts#L96-L140), maybe we can add another emitter like `explorerRootErrorEmitter` but that doesn't look nice either. Any other ideas?